extensions/metadata/interface/src/main/java/org/exist/storage/md/xquery/Reindex.java

```diff
@@ -19,99 +19,87 @@
 import org.exist.xmldb.XmldbURI;
-import org.exist.xquery.BasicFunction;
-import org.exist.xquery.Cardinality;
-import org.exist.xquery.FunctionSignature;
-import org.exist.xquery.XPathException;
-import org.exist.xquery.XQueryContext;
-import org.exist.xquery.value.Sequence;
-import org.exist.xquery.value.SequenceType;
-import org.exist.xquery.value.Type;
+import org.exist.xquery.*;
+import org.exist.xquery.value.*;
 
-import static org.exist.storage.md.MDStorageManager.*;
+import static org.exist.storage.md.MetaData.NAMESPACE_URI;
+import static org.exist.storage.md.MetaData.PREFIX;
+import static org.exist.storage.lock.Lock.READ_LOCK;
 
 /**
+ * 
  * @author <a href="mailto:shabanovd@gmail.com">Dmitriy Shabanov</a>
- *
+ * @author Casey Jordan <casey.jordan@jorsek.com>
  */
 public class Reindex extends BasicFunction {
-	
-	public final static FunctionSignature signature =
+
+	private static final QName NAME = new QName("reindex", NAMESPACE_URI, PREFIX);
+	private static final String DESCRIPTION = "Reindex collection or document.";
+    private static final SequenceType RETURN = new SequenceType(Type.EMPTY, Cardinality.ZERO);
+
+    public final static FunctionSignature signatures[] = {
 		new FunctionSignature(
-			new QName("reindex", NAMESPACE_URI, PREFIX),
-			"",
-			null,
-			new SequenceType(Type.STRING, Cardinality.EMPTY));
+			NAME,
+			DESCRIPTION,
+			new SequenceType[] {
+				 new FunctionParameterSequenceType("resource-url", Type.STRING, Cardinality.ONE_OR_MORE, "The resource's urls.")
+			},
+			RETURN
+		)
+	};
 
 	/**
 	 * @param context
 	 */
-	public Reindex(XQueryContext context) {
+	public Reindex(XQueryContext context, FunctionSignature signature) {
 		super(context, signature);
 	}
 
 	/* (non-Javadoc)
 	 * @see org.exist.xquery.BasicFunction#eval(org.exist.xquery.value.Sequence[], org.exist.xquery.value.Sequence)
 	 */
 	public Sequence eval(Sequence[] args, Sequence contextSequence) throws XPathException {
-		try {
-			final BrokerPool db = BrokerPool.getInstance();
-			try(final DBBroker broker = db.get(Optional.of(db.getSecurityManager().getSystemSubject()))) {
-    			Collection col = broker.getCollection(XmldbURI.ROOT_COLLECTION_URI);
-    			
-    			final MetaData md = MetaData.get();
-    			
-    			checkSub(broker, md, col);
-			}
-		} catch (Exception e) {
-			throw new XPathException(this, e);
-		}
-		
-		return Sequence.EMPTY_SEQUENCE;
-	}
-	
-	private void checkSub(DBBroker broker, MetaData md, Collection col) throws PermissionDeniedException, IOException, LockException, TriggerException {
-		
-        for (Iterator<XmldbURI> i = col.collectionIterator(broker); i.hasNext(); ) {
-            XmldbURI childName = i.next();
-            Collection childColl = broker.getOrCreateCollection(null, XmldbURI.ROOT_COLLECTION_URI.append(childName));
-            
-            checkSub(broker, md, childColl);
+
+        DBBroker broker = getContext().getBroker();
+
+        try (Txn txn = broker.beginTx()) {
```

Comment:
The shortness is not the issue here. I don't want the Broker to have responsibility for Transactions, otherwise the API is unsymmetrical. TransactionManager is responsible for Transaction.